### ISSUE TITLE
Support for child windows? 

### ISSUE THREAD

**Issue by skarnproject** [wangwenx190/framelesshelper#33]
Is there any support for non-top level windows? I specifically mean the widgets that show up in a separate window and QDockWidget titlebar when undocked.

So far the whole library is the **best** out of all frameless QT solutions on Github! Looking great, but would be nice to handle the children as well.

![Screen Shot 2020-11-05 at 11 41 48 AM](https://user-images.githubusercontent.com/10352100/98217713-eaff0a80-1f5b-11eb-828e-999af71ba3c9.png)

**Comment by wangwenx190** [comment]
Child windows may be supported. I just didn't test child windows.

**Comment by wangwenx190** [comment]
By the way, what do you want to tell me by showing this picture?

**Comment by wangwenx190** [comment]
> 
> 
> Child windows may be supported. I just didn't test child windows.

You can try to make the child windows frameless. If it doesn't work, you can tell me and I'll see if there's a good solution.

**Comment by skarnproject** [comment]
Just the fact that it does not affect the child titlebars. (And as I understood from your answer, it should not by default).  There is a QWidget with Tool and StaysOnTopHint in the middle, and two QDockWidgets set floating. For QDockWidget I tried calling FramelessWindowsManager::addWindow(&widget) as well. but it does not work. Should I maybe put a connection to QDockWidget's undock event and only addWindow when undock has occured?

**Comment by wangwenx190** [comment]
>
> Should I maybe put a connection to QDockWidget's undock event and only addWindow when undock has occured?
>
I'm not quite sure about this. Actually, on non-Windows platforms, I'm using the `Qt::FramelessWindowHint` flag to get rid of the title bar. You can try to enable that flag for a normal QDockWidget and a QWidget with Tool and StaysOnTopHint to see if they can become frameless. If they can become frameless indeed, then I need to check my code to see what's wrong.

**Comment by bighzl** [comment]
创建一个非top-level的子window后，原来的top-level window就抓不到WM_NCHITTEST，大概是在
```cpp
const QWindow *window = findWindow(msg->hwnd);
    if (!window || (window && !window->property(m_framelessMode).toBool())) {
        return false;
    }
```
这个地方被返回了。一旦抓不到WM_NCHITTEST事件后，window窗口的title area就开始响应鼠标事件了，这在正常的时候是不会响应的。

**Comment by wangwenx190** [comment]
> 
> 
> 创建一个非top-level的子window后，原来的top-level window就抓不到WM_NCHITTEST，大概是在
> 
> ```c++
> const QWindow *window = findWindow(msg->hwnd);
>     if (!window || (window && !window->property(m_framelessMode).toBool())) {
>         return false;
>     }
> ```
> 
> 这个地方被返回了。一旦抓不到WM_NCHITTEST事件后，window窗口的title area就开始响应鼠标事件了，这在正常的时候是不会响应的。

不是顶层窗口的话就没有`QWindow`，`findWindow`这个函数会返回空指针，在你标出来的这个地方确实会直接返回。

但这里我想要确保`msg->hwnd`这个句柄是一个独立的窗口，而不是一个控件，我暂时不知道有什么更好的方法。用Win32 API的话，我试过`IsWindow`，`GetParent`和`GetAncestor`等函数，都没法很好的区分窗口和控件，而且Qt还会创建一些不可见的窗口来实现一些特别的功能，也会对这些API的效果造成很大的影响。至于Qt自己的API，确实有`isTopLevel`和`isWindow`等函数，但这都是`QWidget`特有的函数，Qt Quick是用不了的，而且还会引入QtWidgets这个依赖项，不太可行。

如果不管`msg->hwnd`是不是一个独立的窗口，直接无脑应用无边框的代码，我也试过，最终效果不好，也不可行。

你知道有什么比较好的办法能判断一个句柄是不是一个独立的窗口吗？

**Comment by Flip20** [comment]
Hi Yuhang Zao,

 

If I try to use the frameless lib in my true app (MainWindow), moving the window by dragging on the custom title bar does not work. I do not receive messages of type WM_NCHITTEST in  FramelessHelperWin::nativeEventFilter, while I do receive other messages there. I’m afraid it has something to do with event filters on my DockWidgets, but not sure. Any idea?

 

Thanks again!

 

Philip

**Comment by Flip20** [comment]
The problem lies in my QVTKWidget2, see also

https://github.com/dfct/TrueFramelessWindow/issues/13

**Comment by wangwenx190** [comment]
@Flip20 Sorry I'm not sure about it either.

**Comment by wangwenx190** [comment]
Should be fixed in 2.0. closing for now. feel free to reopen it if necessary.
